readme.txt
todo : Add proper readme description.
This was so far in the dev branch. Lets merge it to the main branch.


changes.txt
v1: 
added self.qubitCircLeadSizes =  ([2, 20], [10, 100])
      self.qubitCircRadius = 200
as new parameters to the default config class.

edited def qubitLead():

now has a new argument: isCirc. If true, the qubit will be drawn with circular pads, whos radius is specified in self.qubitCircRadius
One thing to look out far is the overlap between the coarseLead and the Circle. 
Currently, the generic value
	self.qubitLeadOverlap =       1.5   
is used.
added snake Resonator.

v1.1: 
changes in the individual pythonfiles. 
Pengtao request the boarder for the fine Lead and corse Lead of the qubit to be in a different Layer. For this,
I adjusted the code. 
First, the boarder of the merged Leads is created to be used as a mask. Then, I create the boarder for both the coarse Lead and fine Lead individually.
Finally, I use the created mask for an and operation on the indiviudall borders.
Also, the createboarder function was adjusted to allow picking a layer. Previously, all borders were in the same layer.

autoLayout.py

"""
    verticalDistances = [        6165,    12845,       800,    750,      685,        685       ]
    horizontalDistances = [3500,     3500,        0,        0,    3500,       415,      3500  ]
"""

def autoLayout(topology, distances):
    horDist, verDist = distances
    parts = []
    sizes = np.zeros([10, 10, 2])
    for i, row in enumerate(topology):
        for j, column in enumerate(row):
            sizes[i,j] = getSize(column)
    for i, row in enumerate(topology):
        for j, column in enumerate(row):
            if len(row)>1:
                dx = (horDist[i] + sizes[i,j,0]) * (j - (len(row)-1)/2)
            else:
                dx = 0
            dy = -( np.sum(verDist[:i]) + np.sum(sizes[:i,0,1]) + sizes[i,j,1] / 2 )
            parts.extend(translate(column, dx, dy))
    return parts

def testAutoLayoutIndexing():
    sizes = [1, 2, 3]
    dist = [4, 5]
    for i, _ in enumerate(sizes):
        print(i, dist[:i], sizes[:i], sizes[i])
testAutoLayoutIndexing()

"""
    topology = [
        [markerL1, markerL2],
        [markerV1, markerV2],
            [resonator1],
              [qubit1],
        [markerL3, markerL4],
        [qubit2, qubit3, qubit4, qubit5],
        [markerL5, markerL6],
    ]
    distances = [ conf.horizontalDistances, conf.verticalDistances ]
"""

ChipBob/ChipBob.py

from QubitEBLDesignV1 import *
from copy import deepcopy
import os
#  I want to save the generated file in the directory of the script.
abspath = os.path.abspath(__file__)
dname = os.path.dirname(abspath)
os.chdir(dname)

def qubit(conf: DefaultConfig, test, isCirc = False, cLlayer = 6, fLlayer = 13 ):
    bridgeFreeJJ1 = bridgeFreeJJ(conf)
    size1 = getSize(bridgeFreeJJ1)
    
    coarseLead1, fineLead1 = translate(qubitLead(conf, test, isCirc), 0, size1[1]/2)
    coarseLead2, fineLead2 = rotate(copy([coarseLead1, fineLead1]), pi, [0,0])
    discharger1 = [] if test else discharger(conf)
    
    # added some mmodifications to allow adjusting fineLead and CoarseLead border layer
    # Create a mask of the joined coarselead + fine lead border to use for boolean and
    _joinedLead = [join([coarseLead1, coarseLead2] + discharger1), fineLead1, fineLead2]
    _border = makeBorder(conf, _joinedLead)
    _border = cut(_border, myRectangle([size1[0]*10, size1[1]]))#.fillet(10, max_points=1000)
    
    joinedcoarseLead = [join([coarseLead1, coarseLead2] + discharger1)]
    joinedfineLead = [fineLead1, fineLead2]
    
    # Create border for coarseLead (large rectangles and Pads)
    bordercL = makeBorder(conf, joinedcoarseLead)
    bordercL = cut(bordercL, myRectangle([size1[0]*10, size1[1]]))#.fillet(10, max_points=1000)
    bordercL = gdspy.boolean(_border,bordercL,"and", layer= cLlayer)
    
    # Create border for fine lead
    borderfL = makeBorder(conf,joinedfineLead)
    borderfL = cut(borderfL, myRectangle([size1[0]*10, size1[1]]))#.fillet(10, max_points=1000)
    borderfL = gdspy.boolean(_border,borderfL,"and", layer= fLlayer)
    borderfL = gdspy.boolean(borderfL,bordercL,"not", layer= fLlayer)
    
    joinedLead = joinedcoarseLead + joinedfineLead
    borders = [bordercL, borderfL]
    return bridgeFreeJJ1 + joinedLead + borders
def filledChip(conf: DefaultConfig, texts):
    verDist, horDist = conf.verticalDistances, conf.horizontalDistances
    # make elements
    markerL1 = markerL(conf, text=texts[0])
    markerL2 = markerL(conf, text=texts[1], rotation= -pi/2)
    markerV1 = markerV(conf)
    markerV2 = markerV(conf, rotation= pi)
    resonator1 = resonator(conf)
    qubit1 = qubit(conf, test=False, isCirc = True)
    snake1 = snake(conf, direction = False)
    markerV3 = markerV(conf, rotation = pi)
    markerV4 = markerV(conf)
    markerV5 = markerV(conf)
    markerV6 = markerV(conf, rotation= pi)
    markerL3, markerL4 = copy(markerL1), copy(markerL2)
    testQubit1 = qubit(conf, test=True, isCirc=True)
    markerL5 = markerL(conf, rotation= pi/2)
    markerL6 = markerL(conf, rotation= pi)
    # calculate positions
    sizes = getSize(resonator1), getSize(qubit1), getSize(snake1), getSize(testQubit1)
    yMarkerV1 = verDist[0]
    yResonator = yMarkerV1 + verDist[1] + sizes[0][1] / 2
    yQubit = yResonator + sizes[0][1] / 2 + verDist[2] + sizes[1][1] / 2
    ySnake = yQubit + sizes[1][1] / 2 + verDist[3] + conf.snakeRadius
    yMarkerV3 = ySnake + sizes[2][1]-conf.snakeRadius + verDist[4]
    yMarkerV5 = yMarkerV3 + verDist[5]
    
    yMarkerL1 = yMarkerV5 + verDist[6]
    yTestQubit = yMarkerL1 + verDist[7] + sizes[3][1] / 2
    yMarkerL2 = yTestQubit + verDist[7] + sizes[3][1] / 2
    xMarker = horDist[0] / 2
    xTestQubit = lambda i: (sizes[3][0] + horDist[1]) * (i- (conf.numTestQubits-1)/2 )
    # translate elements
    translate(markerL1 + markerL3 + markerL5 + markerV1 + markerV3 + markerV5, -xMarker, 0)
    translate(markerL2 + markerL4 + markerL6 + markerV2 + markerV4 + markerV6,  xMarker, 0)
    testQubits = []
    for i in range(conf.numTestQubits):
        testQubits += translate(copy(testQubit1), xTestQubit(i), -yTestQubit) 
    translate(markerV1 + markerV2, 0, -yMarkerV1)
    translate(resonator1         , 0, -yResonator)
    translate(qubit1             , 0, -yQubit)
    translate(snake1             , 0, -ySnake)
    translate(markerV3 + markerV4, 0, -yMarkerV3)
    translate(markerV5 + markerV6, 0, -yMarkerV5)
    translate(markerL3 + markerL4, 0, -yMarkerL1)
    translate(markerL5 + markerL6, 0, -yMarkerL2)
    parts = (
        markerL1 + markerL2 + markerL3 + markerL4 + markerL5 + markerL6
        + markerV1 + markerV2 + resonator1 + qubit1 + snake1 + markerV3 
        + markerV4+ markerV5 + markerV6+ testQubits
    )
    borders = [makeBorder(conf, parts) for parts in [
        markerL1, markerL2, markerL3, markerL4, markerL5, markerL6,
        markerV1, markerV2, resonator1, snake1, markerV3, markerV4, 
        markerV5, markerV6
    ]]
    return parts + [makeChip(conf, parts)] + borders

def filledWafer():
    conf0 = DefaultConfig()
    # modification for Alice
    # resonator
    conf0.resonatorSize = [150, 11600]
    # qubitLead            -- fine --======= corse EEEEEEEE
    conf0.qubitCircRadius = 300
    conf0.qubitCircLeadSizes = ([10, 40], [40, 100])
    conf0.qubitTestSize  =                            80
    conf0.qubitLeadGaps  = (       0,        0         )
    conf0.qubitLeadOverlap =      10      
    # discharger
    conf0.dischargerRadius = 250
    conf0.dischargerWidth = 6
    conf0.dischargerAngles = [pi/2, -pi/2]
    # snake
    conf0.snakeNumHooks = 8
    conf0.snakeFinHorLineLen = 1400
    # vertical distances:      markerL - markerV - resonator - qubit - snake - markerV - markerV - markerL - testQubit
    conf0.verticalDistances = [        2500,    800,       2500,    2200,   1160,      5360,          580,     590,       ]
    # horizontal distances:      marker - marker, testQubit - testQubit
    conf0.horizontalDistances = [      3200,                415         ]
    
    chips = []
    parts = []
    for i, JJWidth in enumerate(np.linspace(1.52, 1.66, 8)):
        confI = deepcopy(conf0)
        confI.bridgeFreeJJSizes[3][0] = JJWidth
        chips.append( filledChip(confI, texts=['Col%d' % (i+1), 'W%.2f' % JJWidth]) )
    size = getSize(chips[0])
    for i, chip in enumerate(chips):
        parts += translate(chip, (size[0] + conf0.chipMargin)*i , 0) 
    moveToOrigin(parts)
    grid = makeGrid(conf0, parts)
    makeFile('ChipBob.gds', parts + grid + wafer(conf0))

filledWafer()
    
    

ChipEve/ChipEve.py

from QubitEBLDesignV1 import *
from copy import deepcopy
import os
#  I want to save the generated file in the directory of the script.
abspath = os.path.abspath(__file__)
dname = os.path.dirname(abspath)
os.chdir(dname)

def qubit(conf: DefaultConfig, test, isCirc = False, cLlayer = 6, fLlayer = 13 ):
    bridgeFreeJJ1 = bridgeFreeJJ(conf)
    size1 = getSize(bridgeFreeJJ1)
    
    coarseLead1, fineLead1 = translate(qubitLead(conf, test, isCirc), 0, size1[1]/2)
    coarseLead2, fineLead2 = rotate(copy([coarseLead1, fineLead1]), pi, [0,0])
    discharger1 = [] if test else discharger(conf)
    
    # added some mmodifications to allow adjusting fineLead and CoarseLead border layer
    # Create a mask of the joined coarselead + fine lead border to use for boolean and
    _joinedLead = [join([coarseLead1, coarseLead2] + discharger1), fineLead1, fineLead2]
    _border = makeBorder(conf, _joinedLead)
    _border = cut(_border, myRectangle([size1[0]*10, size1[1]]))#.fillet(10, max_points=1000)
    
    joinedcoarseLead = [join([coarseLead1, coarseLead2] + discharger1)]
    joinedfineLead = [fineLead1, fineLead2]
    
    # Create border for coarseLead (large rectangles and Pads)
    bordercL = makeBorder(conf, joinedcoarseLead)
    bordercL = cut(bordercL, myRectangle([size1[0]*10, size1[1]]))#.fillet(10, max_points=1000)
    bordercL = gdspy.boolean(_border,bordercL,"and", layer= cLlayer)
    
    # Create border for fine lead
    borderfL = makeBorder(conf,joinedfineLead)
    borderfL = cut(borderfL, myRectangle([size1[0]*10, size1[1]]))#.fillet(10, max_points=1000)
    borderfL = gdspy.boolean(_border,borderfL,"and", layer= fLlayer)
    borderfL = gdspy.boolean(borderfL,bordercL,"not", layer= fLlayer)
    
    joinedLead = joinedcoarseLead + joinedfineLead
    borders = [bordercL, borderfL]
    return bridgeFreeJJ1 + joinedLead + borders

def filledChip(conf: DefaultConfig, texts):
    verDist, horDist = conf.verticalDistances, conf.horizontalDistances
    # make elements
    markerL1 = markerL(conf, text=texts[0])
    markerL2 = markerL(conf, text=texts[1], rotation= -pi/2)
    markerV1 = markerV(conf)
    markerV2 = markerV(conf, rotation= pi)
    resonator1 = resonator(conf)
    qubit1 = qubit(conf, test=False, isCirc = True)
    markerV3 = markerV(conf, rotation = pi)
    markerV4 = markerV(conf)
    markerV5 = markerV(conf)
    markerV6 = markerV(conf, rotation= pi)
    markerL3, markerL4 = copy(markerL1), copy(markerL2)
    testQubit1 = qubit(conf, test=True, isCirc=True)
    markerL5 = markerL(conf, rotation= pi/2)
    markerL6 = markerL(conf, rotation= pi)
    # calculate positions
    sizes =  getSize(qubit1),getSize(resonator1), getSize(testQubit1)
    yMarkerV1 = verDist[0]

    yQubit = yMarkerV1 + sizes[0][1] / 2 + verDist[1] 
    yResonator = yQubit + sizes[0][1] / 2 + verDist[2] + sizes[1][1] / 2
    yMarkerV3 = yResonator+ sizes[1][1]/2 + verDist[3]
    yMarkerV5 = yMarkerV3 + verDist[4]
    yMarkerL1 = yMarkerV5 + verDist[5]
    yTestQubit = yMarkerL1 + verDist[6] + sizes[2][1] / 2
    yMarkerL2 = yTestQubit + verDist[6] + sizes[2][1] / 2
    xMarker = horDist[0] / 2
    xTestQubit = lambda i: (sizes[2][0] + horDist[1]) * (i- (conf.numTestQubits-1)/2 )
    # translate elements
    translate(markerL1 + markerL3 + markerL5 + markerV1 + markerV3 + markerV5, -xMarker, 0)
    translate(markerL2 + markerL4 + markerL6 + markerV2 + markerV4 + markerV6,  xMarker, 0)
    testQubits = []
    for i in range(conf.numTestQubits):
        testQubits += translate(copy(testQubit1), xTestQubit(i), -yTestQubit) 
    translate(markerV1 + markerV2, 0, -yMarkerV1)
    translate(resonator1         , 0, -yResonator)
    translate(qubit1             , 0, -yQubit)
    translate(markerV3 + markerV4, 0, -yMarkerV3)
    translate(markerV5 + markerV6, 0, -yMarkerV5)
    translate(markerL3 + markerL4, 0, -yMarkerL1)
    translate(markerL5 + markerL6, 0, -yMarkerL2)
    parts = (
        markerL1 + markerL2 + markerL3 + markerL4 + markerL5 + markerL6
        + markerV1 + markerV2 + resonator1 + qubit1 +  markerV3 
        + markerV4+ markerV5 + markerV6+ testQubits
    )
    borders = [makeBorder(conf, parts) for parts in [
        markerL1, markerL2, markerL3, markerL4, markerL5, markerL6,
        markerV1, markerV2, resonator1, markerV3, markerV4, 
        markerV5, markerV6
    ]]
    return parts + [makeChip(conf, parts)] + borders

def filledWafer():
    conf0 = DefaultConfig()
    # modification for Alice
    # resonator
    conf0.resonatorSize = [150, 12200]
    # qubitLead            -- fine --======= corse EEEEEEEE
    conf0.qubitCircRadius = 300
    conf0.qubitCircLeadSizes = ([10, 40], [40, 100])
    conf0.qubitTestSize  =                            80
    conf0.qubitLeadGaps  = (       0,        0         )
    conf0.qubitLeadOverlap =      10      
    # discharger
    conf0.dischargerRadius = 250
    conf0.dischargerWidth = 6
    conf0.dischargerAngles = [pi/2, -pi/2]

    # vertical distances:      markerL - markerV - qubit - resonator -  markerV - markerV - markerL - testQubit
    conf0.verticalDistances = [        3000,    4250,       3000,     1055,      5360,     580,     590,       ]
    # horizontal distances:      marker - marker, testQubit - testQubit
    conf0.horizontalDistances = [      3200,                415         ]
    
    chips = []
    parts = []
    for i, JJWidth in enumerate(np.linspace(1.52, 1.66, 8)):
        confI = deepcopy(conf0)
        confI.bridgeFreeJJSizes[3][0] = JJWidth
        chips.append( filledChip(confI, texts=['Col%d' % (i+1), 'W%.2f' % JJWidth]) )
    size = getSize(chips[0])
    for i, chip in enumerate(chips):
        parts += translate(chip, (size[0] + conf0.chipMargin)*i , 0) 
    moveToOrigin(parts)
    grid = makeGrid(conf0, parts)
    makeFile('Eve.gds', parts + grid + wafer(conf0))

filledWafer()
    
    

ChipCharlie/QubitEBLDesignV1.py

"""
conda install -c conda-forge gdspy

The units used to represent shapes in the GDSII format are defined by the user. 
The default unit in gdspy is 1 um. But that can be easily changed by the user.
"""

import gdspy
from gdspy import copy
import numpy as np 
import math

pi = np.pi

class DefaultConfig:
    def __init__(self):
        # wafer
        self.waferRadius = 25397 # radius of the wafer
        self.waferSliceAt = -24122 # position of the cut - the waver left from a vertical line at the specified position is removed
        self.waferLayer = 0
        # grid
        self.gridSize = [290, 300] # [x,y]
        self.gridLayer = 15
        # chip 
        self.chipPadding = 195 # Distances between the outwards facing edges of the markers and the boarders of the respective chip
        self.chipMargin = 100 # distance between two chips
        self.chipLayer = 1
        # marker
        self.markerLayer = 12
        # markerL 
        self.markerLSize = [200, 200]
        self.markerLWidth = 20
        self.markerLTextSize = 40
        # markerV
        self.markerVSize = [400, 200]
        # resonator
        self.resonatorSize = [152, 9143] # resonator dimensions [width, length]
        self.resonatorLayer = 12
        #Snake
        self.snakeRadius = 550 
        self.snakeThickness = 200
        self.snakeNeck = 1100 # Starting at the center of the circle!
        self.snakeCenTurnRad = 200 # CentralTurnRadius
        self.snakeHorLineLen = 1400
        self.snakeNumHooks = 4 # Number of full hooks. A hook is a 180 degree turn and a full line segment
        self.snakeLayer = 12 # set to the same layer as resonator
        self.snakeFinHorLineLen = 700 # len of the Final horiz line. Should be smaller then snake.HorLineLen!
        # qubitLead           -- fine --======= corse EEEEEEEE
        self.qubitLeadSizes = ([2, 20], [10, 100], [400, 800 ]) # fine lead, coarse lead, main body
        self.qubitCircLeadSizes =  ([2, 20], [10, 100])
        self.qubitCircRadius = 200
        self.qubitTestSize  =                             80
        self.qubitLeadGaps  = (       10,        40          )
        self.qubitLeadOverlap =       1.5      
        self.qubitLeadLayers = [  11,            12          ] 
    
        # discharger
        self.dischargerRadius = 400 # seems to be defined as a circle, drawn between the specified angles with the qubit center as its center.
        self.dischargerWidth = 4
        self.dischargerAngles = [pi/2, -pi/2]
        self.dischargerLayer = 12
        # bridgeFreeJJ
        self.bridgeFreeJJSizes = ([0.58, 0.9], [0.5, 0.9], [0.4, 0.7], [1.66, 0.1])
        self.bridgeFreeJJLayers = [8, 10, 9, 7]
        # border
        self.borderWidth = 0.4 # tiny layer sheating every drawn geometry
        self.borderLayer = 6
        # vertical distances:      markerL - markerV - resonator - qubit - markerL - testQubit
        self.verticalDistances = [        6165,    12845,       800,    750,      685,       ]
        # horizontal distances:      marker - marker, testQubit - testQubit
        self.horizontalDistances = [      3500,                415         ]
        # quantities
        self.numTestQubits = 4
        # UNDER TEST
    

# operations

def join(parts): 
    layer = parts[0].layers[0] if isinstance(parts, list) else parts.layers[0]
    return gdspy.boolean(parts,None,'or', layer=layer, max_points=10000)
def cut(a,b): return gdspy.boolean(a,b,'not', layer=a.layers[0], max_points=10000)
def translate(parts, dx, dy): return [a.translate(dx, dy) for a in parts]
def rotate(parts, angle, center): return [a.rotate(angle, center) for a in parts]
def moveToOrigin(parts):
    point1, point2 = join(parts).get_bounding_box()
    dx, dy = -(point1 + point2)/2
    return translate(parts, dx,dy)

def makeBorder(conf: DefaultConfig, parts): 
    # working principle: take an object, expand it, cut the original object from the expanded one and return the result as border.
    return cut(gdspy.offset(parts, conf.borderWidth, layer=conf.borderLayer), parts)
def makeChip(conf: DefaultConfig, parts):
    padding = conf.chipPadding
    point1, point2 = join(parts).get_bounding_box()
    return gdspy.Rectangle(point1 - padding, point2 + padding, layer=conf.chipLayer) 
def makeGrid(conf: DefaultConfig, parts):
    dX, dY = getSize(parts) / 2
    dx, dy = conf.gridSize
    layer = conf.gridLayer
    numX, numY = int(np.ceil(dX/dx)), int(np.ceil(dY/dy))
    rectangles = []
    for i in range(-numX, numX+1):
        for j in range(-numY, numY+1):
            rectangles.append(myRectangle([dx,dy], layer, [i*dx, j*dy]))
    texts = [
        gdspy.Text('Extent %s'     % str([dX, dY]), dy, [-numX*dx, numY*dy + dy*2], layer=layer),
        gdspy.Text('Field Size %s' % str([dx, dy]), dy, [-numX*dx, numY*dy + dy], layer=layer),
    ]
    return rectangles + texts
def makeFile(filename, parts):
    lib = gdspy.GdsLibrary()
    cell = lib.new_cell('cell')
    cell.add(parts)
    cell.add(gdspy.Label('origin', [0,0]))
    lib.write_gds(filename)
    gdspy.LayoutViewer()

# calculations

def getSize(parts): 
    points = join(parts).get_bounding_box()
    return points[1] - points[0]

# basic parts

def myRectangle(size, layer=0, center=[0,0]):
    point1 = np.array(size) * 0.5
    return gdspy.Rectangle(point1, -point1, layer=layer).translate(*center)

# the following functions all returns a list of gdspy part
# parts that appear once

def wafer(conf: DefaultConfig):
    circle = gdspy.Round([0,0], 1, layer=conf.waferLayer,
        tolerance=1e-3).scale(conf.waferRadius)
    wafer = gdspy.slice(circle, conf.waferSliceAt, 0)[1]
    return [wafer]

# parts that appear many times

def markerL(conf: DefaultConfig, text='', rotation=0):
    dx, w, ts = conf.markerLSize[0], conf.markerLWidth, conf.markerLTextSize
    markerL = gdspy.FlexPath([(0,-dx), (0,0), (dx, 0)], w, layer=conf.markerLayer).to_polygonset()
    text = gdspy.Text(text, ts, [w, w], layer=conf.markerLayer).rotate(-pi/2)
    if rotation == -pi/2:
        text.translate(-dx, 0)
    return [markerL.rotate(rotation), text]

def markerV(conf: DefaultConfig, rotation=0):
    dx, dy = conf.markerVSize
    markerV = gdspy.Polygon([(0,0), (dx, -dy/2), (dx, dy/2)], layer=conf.markerLayer)
    return [markerV.rotate(rotation)]

def resonator(conf: DefaultConfig):
    return [myRectangle(conf.resonatorSize, conf.resonatorLayer)]

def snake(conf: DefaultConfig, direction = True): # True means pad at bottom, false, pad at top
    SnakeHead = gdspy.Round([0,0],conf.snakeRadius, layer = conf.snakeLayer)
    SnakeBody = gdspy.Path(conf.snakeThickness, (0, 0))
    # create the neck, first turn and first horizontal line element
    if direction:
        SnakeBody.segment(conf.snakeNeck, "+y")
    else:
        SnakeBody.segment(conf.snakeNeck, "-y")
        
    SnakeBody.turn(conf.snakeCenTurnRad,"r")
    SnakeBody.segment(conf.snakeHorLineLen/2-conf.snakeCenTurnRad)
    
    # create as many full hooks as specified in conf0 plus an additional final turn
    if conf.snakeNumHooks%2 == 0: # if even
        for i in range(math.ceil(conf.snakeNumHooks/2)):
            SnakeBody.turn(conf.snakeCenTurnRad,"ll")
            SnakeBody.segment(conf.snakeHorLineLen)
            SnakeBody.turn(conf.snakeCenTurnRad,"rr")
            SnakeBody.segment(conf.snakeHorLineLen)
        SnakeBody.turn(conf.snakeCenTurnRad,"ll")
    else: # if odd
        SnakeBody.turn(conf.snakeCenTurnRad,"ll")
        SnakeBody.segment(conf.snakeHorLineLen)
        for i in range(math.ceil((conf.snakeNumHooks-1)/2)):
            SnakeBody.turn(conf.snakeCenTurnRad,"rr")
            SnakeBody.segment(conf.snakeHorLineLen)
            SnakeBody.turn(conf.snakeCenTurnRad,"ll")    
            SnakeBody.segment(conf.snakeHorLineLen)
        SnakeBody.turn(conf.snakeCenTurnRad,"rr")        
    # Final Horizontal line segment
    SnakeBody.segment(conf.snakeFinHorLineLen)
    
    Snake = join([SnakeHead] + [SnakeBody])
    return [Snake]

def qubitLead(conf: DefaultConfig, test = False, isCirc = False ): 
    """Create the Qubit Pad, Coarse Lead and Fine Lead. The Pad and Coarse Lead are added together into Corse Lead.

    Args:
        conf (DefaultConfig): dictionary holding all geometric variables
        test (bool, optional): Decides if the Junction is a test junction. Defaults to False.
        isCirc (bool, optional): To turn the qubitPad into a circle. Defaults to False.

    Returns:
        list[polygon, polygon]: [description]
    """
    # Here, I want to either load the dimensions for the circular qubit or the rec one.
    # Also, if it is a test qubit as we mainly care about the junction.
    
    if isCirc:
        sizes = conf.qubitCircLeadSizes
        radius = conf.qubitCircRadius
    else: 
        sizes = conf.qubitLeadSizes
    if test: 
        sizes = conf.qubitLeadSizes
        sizes[2][1] = conf.qubitTestSize
        
    gaps = conf.qubitLeadGaps
    layers = conf.qubitLeadLayers
    curve = gdspy.Curve(0)
    # This part takes care of drawing out the fineLead, CoraseLead and if isCirc = False the pad.
    # The way it works is by drawing half the structure (vertical cut) and then mirroring it over.
    for i, size in enumerate(sizes):
        w1, h1 = size
        if i == 0:
            dx = dy = w1/2
            curve.c(dx,0, dx,0, dx,dy).v(h1-dy)
        else:
            dx, dy = (w1-sizes[i-1][0])/2, gaps[i-1]
            curve.c(0,dy/2, dx,dy/2, dx,dy).v(h1)
    curve.h(-sizes[-1][0]/2)
    # Mirror operation.
    halfLead = gdspy.Polygon(curve.get_points())
    lead = join([ halfLead, copy(halfLead).mirror([0,1]) ])
    # Seperating the single gdspy object into coarse and fine lead
    slice1 = sizes[0][1]+gaps[0]
    slice2 = slice1 + conf.qubitLeadOverlap
    coarseLead = gdspy.slice(lead, slice1, 1, layer=layers)[1]
    fineLead   = gdspy.slice(lead, slice2, 1, layer=layers)[0]
    if isCirc and not test:
        CirclePad = gdspy.Round(center=[0,sizes[0][1]+sizes[1][1]+radius+gaps[0]-conf.qubitLeadOverlap], radius=radius,layer=layers[1])
        coarseLead = join([coarseLead,CirclePad])   
    return coarseLead, fineLead

def discharger(conf: DefaultConfig):
    w, r, angles = conf.dischargerWidth, conf.dischargerRadius, conf.dischargerAngles
    return [gdspy.Path(w, [0, r]).arc(r, *angles, layer=conf.dischargerLayer)]

def bridgeFreeJJ(conf: DefaultConfig):
    sizes = conf.bridgeFreeJJSizes
    layers = conf.bridgeFreeJJLayers
    rectangles = [myRectangle(size, layers[i]) for i,size in enumerate(sizes)]
    dy0 = ( sizes[0][1] + sizes[3][1] ) / 2
    dx1 = ( sizes[0][0] - sizes[1][0] ) / 2
    dx2 = ( sizes[0][0] - sizes[2][0] ) / 2
    rectangles[0].translate(0, dy0)
    rectangles[1].translate(dx1, dy0)
    rectangles[2].translate(dx2, dy0)
    rectangles[0] = cut(rectangles[0], rectangles[1])
    rectangles[1] = cut(rectangles[1], rectangles[2])
    for i in range(3):
        rectangles.append(copy(rectangles[i]).rotate(pi))
    return rectangles
